core/gate.h

```diff
@@ -165,20 +165,39 @@ class Gate {
   DISALLOW_COPY_AND_ASSIGN(Gate);
 };
 
-class IGate;
+class OGate;
+
+// A class for input gate
+class IGate : public Gate {
```

Comment:
I assume you had to move this up so that `OGate::SetIgate` could use `gate_idx()` in line 195? I wonder if it might be worthwhile moving the implementations for IGate and OGate methods to the cc file, since neither is templated or shows any other reason why doing so would be disallowed? This might simplify future changes, but it is entirely optional since I am not sure of the benefits.

Reply:
I agree to the point.  I changed the order of two classes definition because I'm lazy and want to leave a minimal footprint. I will do, it's time to burn my hot chocolate. 🍫